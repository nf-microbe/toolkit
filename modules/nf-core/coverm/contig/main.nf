process COVERM_CONTIG {
    tag "${meta.id}"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/coverm:0.7.0--h07ea13f_1':
        'biocontainers/coverm:0.7.0--h07ea13f_1' }"

    input:
    tuple val(meta) , path(fastq)
    tuple val(meta2), path(fasta)

    output:
    tuple val(meta), path("${prefix}_alignment_results.tsv")    , emit: tsv
    tuple val(meta), path("${prefix}.bam")                      , emit: bam
    tuple val(meta), path("${prefix}.log")                      , emit: log
    path "versions.yml"                                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def reads = !meta.single_end ? "--coupled ${fastq}" : "--single ${fastq}"
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    coverm \\
        contig \\
        ${reads} \\
        --reference ${fasta} \\
        --output-file ${prefix}_alignment_results.tsv \\
        --bam-file-cache-directory ${prefix}_bam_files \\
        --threads ${task.cpus} \\
        ${args} \\
        2> ${prefix}.log

    mv ${prefix}_bam_files/*.bam ${prefix}.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        coverm: \$(echo \$(coverm --version 2>&1) | sed 's/^.*coverm //; s/Using.*\$//' ))
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def reads = !meta.single_end ? "--coupled ${fastq}" : "--single ${fastq}"
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_alignment_results.tsv
    touch ${prefix}.bam
    touch ${prefix}.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        coverm: \$(echo \$(coverm --version 2>&1) | sed 's/^.*coverm //; s/Using.*\$//' ))
    END_VERSIONS
    """
}
