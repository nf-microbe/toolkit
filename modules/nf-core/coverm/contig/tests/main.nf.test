nextflow_process {

    name "Test Process: COVERM_CONTIG"
    script "../main.nf"
    process "COVERM_CONTIG"

    tag "modules"
    tag "modules_nfcore"
    tag "coverm"
    tag "coverm/contig"

    test("fastq.gz se + fasta.gz") {
        when {
            process {
                """
                input[0] = [
                    [ id: 'test1', single_end:true ],
                    [
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz', checkIfExists: true),
                    ]
                ]
                input[1] = [
                    [ id: 'test1' ],
                    file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(
                    process.out.tsv,
                    file(process.out.bam[0][1]).name,
                    file(process.out.log[0][1]).name,
                    process.out.versions
                    ).match()
                }
            )
        }
    }


    test("fastq.gz pe + fasta.gz") {
        when {
            process {
                """
                input[0] = [
                    [ id: 'test1', single_end:false ],
                    [
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test2_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test2_2.fastq.gz', checkIfExists: true)
                    ]
                ]
                input[1] = [
                    [ id: 'test1' ],
                    file("${projectDir}/assets/test-datasets/genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test2.contigs.fa.gz", checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(
                    process.out.tsv,
                    file(process.out.bam[0][1]).name,
                    file(process.out.log[0][1]).name,
                    process.out.versions
                    ).match()
                }
            )
        }
    }


    test("fastq.gz pe + fasta.gz - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id: 'test1' ],
                    [
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz', checkIfExists: true)
                    ]
                ]
                input[1] = [
                    [ id: 'test1' ],
                    file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }
    }
}