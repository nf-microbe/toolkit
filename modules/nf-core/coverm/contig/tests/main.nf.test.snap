{
    "fastq.gz pe + fasta.gz - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1_alignment_results.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,2bede4e7376363a3d53626ecbcac2b8c"
                ],
                "bam": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "log": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "tsv": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1_alignment_results.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2bede4e7376363a3d53626ecbcac2b8c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-07T06:50:21.330486625"
    },
    "fastq.gz pe + fasta.gz": {
        "content": [
            [
                [
                    {
                        "id": "test1",
                        "single_end": false
                    },
                    "test1_alignment_results.tsv:md5,a88de99529660cc909bc0a6b26b65d9d"
                ]
            ],
            "test1.bam",
            "test1.log",
            [
                "versions.yml:md5,2bede4e7376363a3d53626ecbcac2b8c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-07T07:00:09.986489654"
    },
    "fastq.gz se + fasta.gz": {
        "content": [
            [
                [
                    {
                        "id": "test1",
                        "single_end": true
                    },
                    "test1_alignment_results.tsv:md5,a23785959ca535474b095a8511293dc3"
                ]
            ],
            "test1.bam",
            "test1.log",
            [
                "versions.yml:md5,2bede4e7376363a3d53626ecbcac2b8c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T08:41:37.309749488"
    }
}